### PR TITLE
Restructure API / SDK components (Attributes, Clock, Trace)

### PR DESCRIPTION
**Pull request by Nevay**
- moves `Attributes` and `Clock` from SDK/Trace to SDK, needed by metrics too
- moves `Event` and `Link` from API to SDK, only needed for readable spans, not used by API
- moves `NoopTracer` from SDK to API as required by specification
- moves `AttributesInterface` from API to SDK by widening types in API to iterable
  previous implementation required API consumers to depend on the SDK implementation to provide attributes to API methods
  bc-breaks:
  - `Attribute` abstraction is removed, `Attributes` now return only the value instead of {name,value} pairs
  - `iterable $attributes` arguments are now non-nullable
  - (todo: rename `Attributes::setAttribute()` to `Attributes::set()` to be consistent with `Attributes::get()` ?)
- moves `ClockInterface` from API to SDK
  currently only used in SDK, interface+implementation should either be fully available in API or just in SDK

### CODE REVIEW

File: src/SDK/Trace/ImmutableSpan.php [hunk 1]
```diff
@@ -131,7 +132,7 @@ public function getEvents(): array
         return $this->events;
     }
 
-    public function getAttributes(): API\AttributesInterface
+    public function getAttributes(): AttributesInterface
```
Comment (brettmc) [inline src/SDK/Trace/ImmutableSpan.php:135]: I'm a little confused about the mixed usage of `iterable` and `AttributesInterface` for attributes in different parts of the code. I just want to make sure that's intentional?

File: tests/SDK/Util/SpanData.php [hunk 2]
```diff
@@ -10,20 +10,23 @@
...
+    /** @var list<EventInterface> */
     private array $events = [];
 
-    /** @var list<API\LinkInterface> */
+    /** @var list<LinkInterface>
+     */
     private array $links = [];
 
-    private API\AttributesInterface $attributes;
+    private \OpenTelemetry\SDK\AttributesInterface $attributes;
```
Comment (brettmc) [inline tests/SDK/Util/SpanData.php:29]: I think the project style is to have these as `use` statements

File: src/SDK/Trace/ImmutableSpan.php (hunk 1 above)
Comment (Nevay) [reply src/SDK/Trace/ImmutableSpan.php:135]: Its intentional, API uses `iterable` (only parameters), SDK uses `AttributesInterface`.

File: tests/SDK/Util/SpanData.php (hunk 2 above)
Comment (Nevay) [reply tests/SDK/Util/SpanData.php:29]: Missed this one, PhpStorm changed some use-statements to fqcn on moving these classes.

**Comment by brettmc** [comment]
I've taken a first pass over this - I like! Couple of comments to be looked at, and I'm in favour of renaming `setAttribute` to `set` - we're still pre-alpha so breaking things for the better is fine.

**Review by tidal — CHANGES_REQUESTED** [review]

**Review by tidal — APPROVED** [review]

**Comment by tidal** [comment]
Thanks for your changes (sorry for the delay, I was busy over the holidays).
There are some merge conflics now, otherwise It's looking good imo.

**Comment by brettmc** [comment]
Hi @Nevay - are you happy for this to be merged now?

**Comment by Nevay** [comment]
@brettmc yes, this should be finished now.
I kept `Attributes::setAttribute()` unchanged for now - method will be removed when we split `Attributes` into mutable builder and immutable attributes to ensure that readonly objects (e.g. `SpanDataInterface`) are not indirectly mutable.

**Pull request merged**